draft-ietf-dnsop-svcb-https.md

```diff
@@ -659,6 +659,35 @@ each RRSet in the Additional section with the same DNSSEC-related records
 that they would send when providing that RRSet as an Answer (e.g. RRSIG, NSEC,
 NSEC3).
 
+## EDNS Client Subnet (ECS)
+
+The EDNS Client Subnet extension (ECS, {{!RFC7871}}) allows recursive
+resolvers to request IP addresses that are suitable for a particular client
+IP range.  IP addresses appear in two places in Authoritative SVCB responses:
+
+* ipv*hint SvcParams in a ServiceMode response.  Clients only use these hints
+when the recursive resolver does not provide A/AAAA records for TargetName in
+the Additional section, as recommended in {{recursive-behavior}}.
+* A/AAAA records returned by the Authoritative server in the Additional
+section of a SVCB response when TargetName is in-zone, as recommended in
+{{authoritative-behavior}}.
+
+SVCB records SHOULD NOT otherwise vary based on ECS.
```

Comment:
Would it be acceptable to encode geo information into a TargetName based on ECS?  Or to not include QUIC in ASNs where QUIC is known to be broken?  What is the rationale for this SHOULD NOT?

Reply:
The behavior described here means that many (most?) ECS-capable resolvers will send "/0" in SVCB queries, so any attempt to make use of ECS for those features will fail.

Also, the majority of resolvers simply don't do ECS at all, so ECS generally is not usable for anything other than performance optimizations.

Reply:
What is the reasoning here? The HTTPSSVC record has great potentials for DNS steering with the use of the priority field to favor closer location while providing some higher latency fallback locations.

Defining per qtype behavior for ECS will also add complexity and is most likely unexpected by most auth DNS operators.

Reply:
DNS steering with priority is allowed within this rule.  It would look like
```dns
; static
www.example.com HTTPS IN 600 1 .
                             2 fallback.example.com.
; ECS-controlled
www.example.com AAAA ...
fallback.example.com AAAA ...
```

From the auth side, the per-QTYPE behavior here is only for authoritative servers that already have per-QTYPE behavior for SVCB (appending the Additional A/AAA records).  Arguably, this is not actually per-QTYPE behavior: the same logic applies to MX and SRV, if anyone wanted to use ECS with them, because they do the same trick of putting A/AAAA in Additional, and Additional can't use ECS.

Reply:
For what it's worth, there are also plenty of places where ECS will get used today
on CNAME records that end up pointing to SVCB records.  For example
the following will be part of a very common deployment path that is widely used today:

```
www.foo.example. CNAME IN 600 foo.georedir.example.net.
foo.georedir.example.net CNAME IN 600 foo.cdn.example.
foo.cdn.example. HTTPS IN 600 ...
foo.cdn.example. A IN 600 ...
foo.cdn.example. AAAA IN 600 ...
```

In this case, "foo.georedir.example.net." may be doing multi-CDN splitting
based on client geo/asn and using ECS as part of this.  
As such, having a "SHOULD NOT" for ECS with SVCB seems likely
to be problematic since the authority here issuing the SVCB has no control
over whether a CNAME pointing to it is using ECS.

Just changing the A/AAAA case won't help when the protocols may need 
to vary (eg, "ISP ___ in the US breaks QUIC by randomly dropping UDP
and thus we don't want to offer it for clients in their ASN").

Reply:
> DNS steering with priority is allowed within this rule. It would look like
> 
> ```
> ; static
> www.example.com HTTPS IN 600 1 .
>                              2 fallback.example.com.
> ; ECS-controlled
> www.example.com AAAA ...
> fallback.example.com AAAA ...
> ```

That will requires two round trips from the resolver to the auth, assuming the resolver supports filling the iphints or worse, RTT from the client when resolver support is missing. We are losing a latency improvement opportunity here, where the auth could just provide a self contained response to minimize traffic and latency while doing steering. 

I think iphint from auth is especially important in the context of DNS steering as those A/AAAA have limited TTL and thus, those additional round trips won't be the exception.

To me the only reason for adding this clause is economical on the resolver side. Having a new popular qtype with ECS cache split can be expensive, especially for a huge resolver like quad8. If that's the motivation, it should be stated clearly, as I see no purely technical reason for this change.

Reply:
@enygren
> For what it's worth, there are also plenty of places where ECS will get used today on CNAME records

I think ECS-variant CNAMEs should probably have been banned in the ECS spec, and this seems like a good opportunity to start phasing them out.  However, if we decide that they should be well-supported, then you're right that we need ECS on the SVCB queries.  I've drawn up #309 to show what I think that would look like.

> Just changing the A/AAAA case won't help when the protocols may need
> to vary (eg, "ISP ___ in the US breaks QUIC by randomly dropping UDP
> and thus we don't want to offer it for clients in their ASN").

Yuck.  You can't rely on ECS, so clients need to handle this anyway.  Any network of this kind is horribly broken, so we shouldn't be trying to put in performance optimizations for it.

@rs
> That will requires two round trips from the resolver to the auth

The number of roundtrips actually has nothing to do with whether the resolver is allowed to disable ECS.  I made a slight text update that I hope makes the reasoning clearer; you can also look at #309.

> To me the only reason for adding this clause is economical on the resolver side. Having a new popular qtype with ECS cache split can be expensive

Yes, the motivation is cache performance, but the resolver can operate with any cache size it likes, so it's not exactly an "expense".  Even with an infinite cache size, unnecessary use of ECS harms performance for users, by making cache hits less likely.

Reply:
> @enygren
> 
> > For what it's worth, there are also plenty of places where ECS will get used today on CNAME records
> 
> I think ECS-variant CNAMEs should probably have been banned in the ECS spec, and this seems like a good opportunity to start phasing them out. However, if we decide that they should be well-supported, then you're right that we need ECS on the SVCB queries. I've drawn up #309 to show what I think that would look like.

FWIW CNAME processing is mentioned in RFC 7871 (section 7.2.1). Quote:
   Since some queries can result in multiple RRsets being added to the
   response, there is an unfortunate ambiguity from the original
   specification as to how SCOPE PREFIX-LENGTH would apply to each
   individual RRset.  For example, multiple types in response to an ANY
   metaquery could all have different applicable SCOPE PREFIX-LENGTH
   values, but this protocol only has the ability to signal one.  The
   response SHOULD therefore, include the longest relevant PREFIX-LENGTH
   of any RRset in the answer, which could have the unfortunate side
   effect of redundantly caching some data that could be cached more
   broadly.  For the specific case of a Canonical Name (CNAME) chain,
   the Authoritative Nameserver SHOULD only place the initial CNAME
   record in the Answer section, to have it cached unambiguously and
   appropriately. Most modern Recursive Resolvers restart the query
   with the CNAME, so the remainder of the chain is typically ignored
   anyway.  For message-focused resolvers, rather than RRset-focused
   ones, this will mean caching the entire CNAME chain at the longest
   PREFIX-LENGTH of any RRset in the chain.

There is no mention of the scope to use for the CNAME response which would imply a non-zero scope is allowed. It does however recommend splitting the response across CNAME, requested RR if the prefix-length scope is different (especially if the CNAME scope is /0) for more efficient caching. It will be good to have reference to this section and a similar recommendation to split CNAME, SVCB alias, SVCB target RRs with different ECS scopes.

> 
> > Just changing the A/AAAA case won't help when the protocols may need
> > to vary (eg, "ISP ___ in the US breaks QUIC by randomly dropping UDP
> > and thus we don't want to offer it for clients in their ASN").
> 
> Yuck. You can't rely on ECS, so clients need to handle this anyway. Any network of this kind is horribly broken, so we shouldn't be trying to put in performance optimizations for it.
> 
> @rs
> 
> > That will requires two round trips from the resolver to the auth
> 
> The number of roundtrips actually has nothing to do with whether the resolver is allowed to disable ECS. I made a slight text update that I hope makes the reasoning clearer; you can also look at #309.
> 
> > To me the only reason for adding this clause is economical on the resolver side. Having a new popular qtype with ECS cache split can be expensive
> 
> Yes, the motivation is cache performance, but the resolver can operate with any cache size it likes, so it's not exactly an "expense". Even with an infinite cache size, unnecessary use of ECS harms performance for users, by making cache hits less likely.

Reply:
Suggested wording to put the ECS RFC text first and the recommendations for auth, recursive after that. Keeping this here instead of in #309 since all the discussion is here.

```suggestion
SVCB records may contain IP addresses (in ipv*hint SvcParams),
or direct users to a TargetName. In these cases, a recursive
resolver and authoritative server can use ECS [expand] to request and
receive answers suitable for a particular client IP range.

According to Section 7.3.1 of RFC7871, "Any records from \[the
Additional section\] MUST NOT be tied to a network". In addition,
Section 7.2.1 states "For the specific case of a Canonical Name (CNAME) chain,
the Authoritative Nameserver SHOULD only place the initial CNAME
record in the Answer section, to have it cached unambiguously and
appropriately.". This is meant to allow caching different RRs at the
appropriate scope prefix-length by a resolver.

Accordingly, authoritative servers SHOULD not include additional RRs
in a response to an HTTPS query if their prefix-length scope is
different from the first RR in the response (which could be a CNAME or an SVCB record). A recursive resolver SHOULD include the same ECS option in SVCB queries as in A/AAAA queries.
```

Reply:
Thanks for the suggestion.  I've incorporated these references in https://github.com/MikeBishop/dns-alt-svc/pull/308/commits/9e0fe746a458fdd2bf7f63d0c357009efcfdb428.